### DISCUSSION TITLE
Does x-bind not work in CSP mode?

### DISCUSSION THREAD

**Discussion by kg-currenxie** [alpinejs/alpine#4049] [7. Bugs (archived)] [answered]
Hi. The CSP example shows the `x-data` example and how to move it out to `Alpine.data()`.

But it seems that any x-bind directive also don't work.

> Since Alpine can no longer interpret strings as plain JavaScrip

Because of this, correct?

```
Alpine.data('sidebar', () => {
  return {
    isLeftSidebarOpen: false
  }
})

<div x-data="sidebar">
  <p x-text="isLeftSidebarOpen" :class="leftSidebarClass ? 'text-primary' : 'text-on-surface-secondary'"></p>
</div>
```

something as simple as this to toggle classes doesn't work. that's expected/limitation?

---

Something crazy like this would work:
```
Alpine.data('sidebar', () => {
  return {
    isLeftSidebarOpen: false,
    isRightSidebarOpen: false,
    leftSidebarClass: '',
    toggleLeftSidebar: function (value) {
      this.isLeftSidebarOpen = !this.isLeftSidebarOpen
      this.leftSidebarClass = this.isLeftSidebarOpen
        ? 'text-primary'
        : 'text-secondary'
      this.$watch('isLeftSidebarOpen', (value) => {
        return (this.leftSidebarClass = value
          ? 'text-primary'
          : 'text-secondary')
      })
    },
  }
})

<p x-text="isLeftSidebarOpen" :class="leftSidebarClass"></p>
```

but the whole point of Alpine is basically gone now ..

or am i doing something wrong?

> @alpinejs/csp v3.13.5

**Comment by ekwoka** [comment, marked as answer]
in CSP mode, expressions are not evaluated at all (may need to check if there is some small parsing). You cannot just put expressions in the directives. Instead they must be the string names of exact things on the component context.

CSP does not allow the use of the function constructor which is necessary to evaluate expressions.

But to do this best in CSP mode would be more like

```js
Alpine.data('sidebar', () => ({
  isLeftSidebarOpen: false,
  get leftSideClass() {
    return this.isLeftSidebarOpen
      ? 'text-primary'
      : 'text-secondary';
  },
  toggleLeftSidebar() {
    this.isLeftSidebarOpen = !this.isLeftSidebarOpen;
  },
}));
```
```html
<p x-text="isLeftSidebarOpen" :class="leftSidebarClass"></p>
```

as demonstrated here: [Alpine Play Sandbox](https://awesomealpine.com/play?coreplugins=0&html=PGRpdiB4LWRhdGE9InNpZGViYXIiPgogIDxidXR0b24KICAgIHR5cGU9ImJ1dHRvbiIKICAgIHgtdGV4dD0iaXNMZWZ0U2lkZWJhck9wZW4iCiAgICA6Y2xhc3M9ImxlZnRTaWRlQ2xhc3MiCiAgICBAY2xpY2s9InRvZ2dsZUxlZnRTaWRlYmFyIgogID48L2J1dHRvbj4KPC9kaXY-Cgo8c3R5bGU-CiAgLnRleHQtcHJpbWFyeSB7CiAgICBjb2xvcjogd2hpdGU7CiAgICBiYWNrZ3JvdW5kLWNvbG9yOiBibGFjazsKICB9CiAgLnRleHQtc2Vjb25kYXJ5IHsKICAgIGNvbG9yOiBibGFjazsKICAgIGJhY2tncm91bmQtY29sb3I6IHdoaXRlCiAgfQo8L3N0eWxlPgo&script=QWxwaW5lLmRhdGEoJ3NpZGViYXInLCAoKSA9PiAoewogIGlzTGVmdFNpZGViYXJPcGVuOiBmYWxzZSwKICBnZXQgbGVmdFNpZGVDbGFzcygpIHsKICAgIHJldHVybiB0aGlzLmlzTGVmdFNpZGViYXJPcGVuCiAgICAgID8gJ3RleHQtcHJpbWFyeScKICAgICAgOiAndGV4dC1zZWNvbmRhcnknOwogIH0sCiAgdG9nZ2xlTGVmdFNpZGViYXIoKSB7CiAgICB0aGlzLmlzTGVmdFNpZGViYXJPcGVuID0gIXRoaXMuaXNMZWZ0U2lkZWJhck9wZW47CiAgfSwKfSkpOwo&panel=script)

Yes, CSP fundamentally reduces how much niceties Alpine provides, but it does still allow clean in DOM data and event binding.

It's possible that future improvements to the CSP evaluator could cover more common use case JS expressions.

**Comment by kg-currenxie** [comment]
For anyone coming here later; I came up with this, which is at least more re-usable for common scenarios (when your state is a boolean for example)

```
Alpine.data('toggle', () => {
  return {
    isToggled: false,
    onToggle() {
      this.isToggled = !this.isToggled
    },
    onOpen() {
      this.isToggled = true
    },
    onClose() {
      this.isToggled = false
    },
    get toggledClass() {
      const appliedClass = this.isToggled
        ? this.$el.getAttribute('data-bound-class')
        : this.$el.getAttribute('data-unbound-class')

      return appliedClass
    },
  }
})
```

and in the template:
```
<div x-data="toggle">
  <div :class="toggledClass" data-bound-class="block" data-unbound-class="hidden"><div>
</div>
```

at least, we can keep the classes in the template instead of in js-files

**Reply by ekwoka** [reply]
could similarly use an attribute like `open` and bind styling to that attribute.

**Reply by kg-currenxie** [reply]
just an example :) not rly using block/hidden

for example

```
data-bound-class="opacity-100 -translate-x-0"
data-unbound-class="opacity-0 -translate-x-full"
```
